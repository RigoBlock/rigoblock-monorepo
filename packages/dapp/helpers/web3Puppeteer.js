const Helper = codecept_helper
const fs = require('fs-extra')
const ganache = require('ganache-cli')
const mnemonic = require('../package.json').config.mnemonic
const pushAction = require('react-router-redux').CALL_HISTORY_METHOD
const fetch = require('node-fetch')

class Web3Puppeteer extends Helper {
  constructor() {
    super()
    this.ganache = null
    this.store = null
  }

  async _before() {
    fs.emptyDirSync('./ganache-db')
    console.log('Database directory emptied.')
    fs.copySync('./ganache-backup', './ganache-db')
    console.log('Ganache backup restored')
    this.ganache = ganache.server({
      mnemonic,
      db_path: './ganache-db',
      port: 8545,
      network_id: 5777
    })
    this.ganache.listen(
      8545,
      err =>
        err
          ? console.error('Error occurred during Ganache startup', err)
          : console.log('Ganache starting!')
    )
    await this.helpers['Puppeteer'].amOnPage('/login')
    await this.inject()
    await this.helpers['Puppeteer'].page.waitForNavigation()
  }

  async _after() {
    await new Promise((resolve, reject) => {
      this.ganache.close(
        err =>
          err
            ? reject(new Error(err))
            : resolve(console.log('Ganache stopping...'))
      )
    })
    this.ganache = null
  }

  async inject() {
    const web3Raw = await fetch(
      'https://cdn.jsdelivr.net/npm/web3@0.20.6/dist/web3.min.js'
    ).then(res => res.text())
    const page = this.helpers['Puppeteer'].page
    await page.evaluate(web3Raw => {
      eval(web3Raw)
      window.web3 = new window.Web3(
        new Web3.providers.HttpProvider('http://localhost:8545')
      )
      this.store = window.init()
    }, web3Raw)
  }

  async navigateToUrl(url) {
    const page = this.helpers['Puppeteer'].page
    await page.evaluate(
      (url, pushAction) =>
        this.store.dispatch({
          type: pushAction,
          payload: { method: 'push', args: [url] }
        }),
      url,
      pushAction
    )
  }
}

module.exports = Web3Puppeteer
