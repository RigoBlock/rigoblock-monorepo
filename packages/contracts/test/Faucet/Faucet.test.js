import { BigNumber } from 'bignumber.js'
import { duration, increaseTimeTo } from '../helpers/increaseTime'
import { latestTime } from '../helpers/latestTime'

const contractName = 'Faucet'
const balance = 2000000000000000000000
const amount = 3000000000000000000000
const oneToken = 1000000000000000000
let faucet
let rigoToken

describeContract(contractName, () => {
  beforeAll(async () => {
    faucet = baseContracts['Faucet']
    rigoToken = baseContracts['RigoToken']
  })

  describe('Faucet Contract - Deploys Correctly', () => {
    const faucet_name = 'GRGFaucet'

    it('Correctly sets Faucet owner', async () => {
      const ownerAddress = await faucet.owner()
      expect(ownerAddress).toBe(accounts[0])
    })

    it('Correctly sets Faucet contract name', async () => {
      const faucetName = await faucet.faucetName()
      expect(faucetName).toBe(faucet_name)
    })

    it('Correctly sets Faucet GRG token address', async () => {
      const faucetInstance = await faucet.tokenInstance()
      expect(faucetInstance).toBe(rigoToken.address)
    })

    it('Correctly sets Faucet status to true after contract instance', async () => {
      const faucetStatus = await faucet.faucetStatus()
      expect(faucetStatus).toBe(true)
    })
  })

  describe('Faucet Contract - drips 1 Token', () => {
    it('Correctly drips 1 Token', async () => {
      // Donor sends GRG to faucet
      await rigoToken.transfer(faucet.address, amount)
      let faucetBalance = await rigoToken.balanceOf(faucet.address)
      expect(new BigNumber(faucetBalance).toFixed()).toBe(
        new BigNumber(amount).toFixed()
      )

      // Faucet drips 1 token
      await faucet.drip1Token.sendTransactionAsync({
        from: accounts[1]
      })
      let dripAccountBalance = await rigoToken.balanceOf(accounts[1])
      expect(new BigNumber(dripAccountBalance).toFixed()).toBe(
        new BigNumber(oneToken).toFixed()
      )
      let faucetBalanceDrip = await rigoToken.balanceOf(faucet.address)
      expect(new BigNumber(faucetBalanceDrip).toFixed()).toBe(
        new BigNumber(faucetBalance).minus(oneToken).toFixed()
      )

      //check previous sender is now locked from requesting again
      await expect(
        faucet.drip1Token.sendTransactionAsync({
          from: accounts[1]
        })
      ).rejects.toThrowErrorMatchingSnapshot()

      //advancing 48 hour in time
      let lastBlockTimestamp = await latestTime()
      const unlockTime = lastBlockTimestamp + duration.hours(48)
      await increaseTimeTo(unlockTime)

      //previous sender can now request again
      await faucet.drip1Token.sendTransactionAsync({
        from: accounts[1]
      })
      dripAccountBalance = await rigoToken.balanceOf(accounts[1])
      expect(new BigNumber(dripAccountBalance).toFixed()).toBe(
        new BigNumber(oneToken).multipliedBy(2).toFixed()
      )
      faucetBalanceDrip = await rigoToken.balanceOf(faucet.address)
      expect(new BigNumber(faucetBalanceDrip).toFixed()).toBe(
        new BigNumber(faucetBalance).minus(oneToken * 2).toFixed()
      )
    })
  })

  describe('Faucet Contract - Faucet OFF and ON', () => {
    it('Correctly turns Faucet OFF and ON', async () => {
      // Turn faucet OFF
      await faucet.turnFaucetOff.sendTransactionAsync({
        from: accounts[0]
      })

      let faucetStatus = await faucet.faucetStatus()
      expect(faucetStatus).toBe(false)

      // Cannot turn faucet OFF again
      await expect(
        faucet.turnFaucetOff.sendTransactionAsync({
          from: accounts[0]
        })
      ).rejects.toThrowErrorMatchingSnapshot()

      faucetStatus = await faucet.faucetStatus()
      expect(faucetStatus).toBe(false)

      //Donor sends GRG to faucet
      await rigoToken.transfer(faucet.address, balance)

      // Cannot drip if faucet is off
      await expect(
        faucet.drip1Token.sendTransactionAsync({
          from: accounts[0]
        })
      ).rejects.toThrowErrorMatchingSnapshot()

      // Turn faucet ON
      await faucet.turnFaucetOn.sendTransactionAsync({
        from: accounts[0]
      })

      faucetStatus = await faucet.faucetStatus()
      expect(faucetStatus).toBe(true)

      // Cannot turn faucet ON again
      await expect(
        faucet.turnFaucetOn.sendTransactionAsync({
          from: accounts[0]
        })
      ).rejects.toThrowErrorMatchingSnapshot()

      faucetStatus = await faucet.faucetStatus()
      expect(faucetStatus).toBe(true)
    })

    it('Non owner cannot turn faucet OFF/ON', async () => {
      // Non owner cannot turn OFF
      await expect(
        faucet.turnFaucetOff.sendTransactionAsync({
          from: accounts[1]
        })
      ).rejects.toThrowErrorMatchingSnapshot()

      // Non owner cannot turn ON
      await expect(
        faucet.turnFaucetOff.sendTransactionAsync({
          from: accounts[1]
        })
      ).rejects.toThrowErrorMatchingSnapshot()
    })
  })
})
