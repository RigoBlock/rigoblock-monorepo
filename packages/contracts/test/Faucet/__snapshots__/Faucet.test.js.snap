// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Faucet Faucet Contract - Faucet OFF and ON Correctly turns Faucet OFF and ON 1`] = `"VM Exception while processing transaction: revert Faucet has to be on"`;

exports[` Faucet Faucet Contract - Faucet OFF and ON Correctly turns Faucet OFF and ON 2`] = `"VM Exception while processing transaction: revert Faucet has to be on"`;

exports[` Faucet Faucet Contract - Faucet OFF and ON Correctly turns Faucet OFF and ON 3`] = `"VM Exception while processing transaction: revert Faucet has to be off"`;

exports[` Faucet Faucet Contract - Faucet OFF and ON Non owner cannot turn faucet OFF/ON 1`] = `"VM Exception while processing transaction: revert"`;

exports[` Faucet Faucet Contract - Faucet OFF and ON Non owner cannot turn faucet OFF/ON 2`] = `"VM Exception while processing transaction: revert"`;

exports[` Faucet Faucet Contract - drips 1 Token Correctly drips 1 Token 1`] = `"VM Exception while processing transaction: revert Required too early."`;

exports[` Faucet Faucet Contract - receive and withdraw Correctly receive and withdraw tokens 1`] = `"VM Exception while processing transaction: revert"`;
